### PR TITLE
Feature/explicit anchor on import

### PR DESCRIPTION
**Pull request by LNHarders**
First step for https://github.com/ampproject/amp.dev/issues/2370
Explicit anchors in Github notation are added on import

### CODE REVIEW

File: platform/lib/pipeline/markdownDocument.js [hunk 1]
```diff
@@ -334,6 +334,32 @@ class MarkdownDocument {
     return true;
   }
 
+  /**
+   *Adds explicit anchors for titels in github notation
+   *
```
Comment (sebastianbenz) [inline platform/lib/pipeline/markdownDocument.js]: nit: remove empty line

File: platform/lib/pipeline/markdownDocument.js [hunk 2]
```diff
@@ -334,6 +334,32 @@ class MarkdownDocument {
...
+   */
+  addExplicitAnchors() {
+    const TITLE_PATTERN = /^#+[ \t]+(?!.*<a[ \t]+name)(.+?)[ \t]*$/mg;
```
Comment (sebastianbenz) [inline platform/lib/pipeline/markdownDocument.js]: please move constants to the top of then file

File: platform/lib/pipeline/markdownDocument.js [hunk 3]
```diff
@@ -334,6 +334,32 @@ class MarkdownDocument {
...
+    const existingSlugs = [];
+    this._contents = this._contents.replace(TITLE_PATTERN, (line, headline) => {
+      let slug = headline.toLowerCase();
+      slug = slug.replace(/ /g, '-');
+      slug = slug.replace(/[^\p{L}0-9_-]/gu, '');
+      if (existingSlugs.includes(slug)) {
```
Comment (sebastianbenz) [inline platform/lib/pipeline/markdownDocument.js]: a for loop avoids duplicating the condition in if and while.

File: platform/lib/pipeline/markdownDocument.js [hunk 4]
```diff
@@ -334,6 +334,32 @@ class MarkdownDocument {
...
 
+  /**
+   *Adds explicit anchors for titels in github notation
+   *
+   */
+  addExplicitAnchors() {
+    const TITLE_PATTERN = /^#+[ \t]+(?!.*<a[ \t]+name)(.+?)[ \t]*$/mg;
+    const existingSlugs = [];
+    this._contents = this._contents.replace(TITLE_PATTERN, (line, headline) => {
+      let slug = headline.toLowerCase();
```
Comment (sebastianbenz) [inline platform/lib/pipeline/markdownDocument.js]: please extract the slug creation logic into a helper class (ideally outside markdown document).

**Comment by LNHarders** [comment]
Implemented the requested changes please check again

File: platform/lib/pipeline/markdownDocument.js [hunk 5]
```diff
@@ -76,6 +77,9 @@ const RELATIVE_LINK_PATTERN = new RegExp(
     /\[[^\]]+\]\(([^:\)\{?#]+)(?:[?#][^\)]*)?\)/.source
     , 'g');
 
+// This pattern will find markdown titels without HTML anchors.
```
Comment (sebastianbenz) [inline platform/lib/pipeline/markdownDocument.js]: s/titels/titles

File: platform/lib/utils/slugGenerator.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Generates unique slugs like github does for anchors
+ */
+class SlugGenerator {
+  constructor() {
+    this.existingSlugs = [];
+  }
+
+  getSlug(headline) {
```
Comment (sebastianbenz) [inline platform/lib/utils/slugGenerator.js]: nit: rename to `generateSlug`. `get*` implies returning an existing value.

**Comment by LNHarders** [comment]
Thanks for the feedback! Please check again.

**Pull request merged**